File: shipley_checks/src/Stage_4/guild_scorer_rust/src/explanation/pest_analysis.rs
use polars::prelude::*;
use anyhow::Result;
use serde::{Serialize, Deserialize};
use rustc_hash::FxHashMap;
use crate::explanation::unified_taxonomy::{OrganismCategory, OrganismRole};
use crate::utils::get_display_name;

/// Pest profile for a guild (qualitative information)
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PestProfile {
    pub total_unique_pests: usize,
    pub top_pests: Vec<TopPest>,
    pub vulnerable_plants: Vec<VulnerablePlant>,
}

/// Top pest by total interactions
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TopPest {
    pub pest_name: String,
    pub category: String,
    pub plant_count: usize,
    pub plants: Vec<String>,
}

/// Plant with its pest vulnerability
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct VulnerablePlant {
    pub plant_name: String,
    pub pest_count: usize,
}

/// Analyze pest profile for a guild
///
/// Extracts herbivore information and identifies:
/// - Shared pests (generalists attacking 2+ plants)
/// - Top 10 pests by interaction count
/// - Most vulnerable plants
pub fn analyze_guild_pests(
    guild_plants: &DataFrame,
    organism_categories: &FxHashMap<String, String>,
) -> Result<Option<PestProfile>> {
    // Check if herbivores column exists (try both list and string formats)
    let herbivores_col = match guild_plants.column("herbivores") {
        Ok(col) => col,
        Err(_) => return Ok(None),
    };

    // Try both list format (Phase 0-4) and string format (legacy)
    let herbivores_list_col = herbivores_col.list().ok();
    let herbivores_str_col = herbivores_col.str().ok();

    let plant_names = guild_plants.column("wfo_taxon_name")?.str()?;

    // Try to get vernacular name columns for display formatting
    let vernacular_en = guild_plants.column("vernacular_name_en").ok().and_then(|c| c.str().ok());
    let vernacular_zh = guild_plants.column("vernacular_name_zh").ok().and_then(|c| c.str().ok());

    // Build pest-to-plants mapping with display names
    let mut pest_to_plants: FxHashMap<String, Vec<String>> = FxHashMap::default();
    let mut plant_pest_counts: FxHashMap<String, usize> = FxHashMap::default();

    for idx in 0..guild_plants.height() {
        if let Some(plant_name) = plant_names.get(idx) {
            let scientific_name = plant_name.to_string();

            // Build display name with vernacular (e.g., "Prunus spinosa (Blackthorn)")
            let en = vernacular_en.and_then(|col| col.get(idx));
            let zh = vernacular_zh.and_then(|col| col.get(idx));
            let display_name = get_display_name(&scientific_name, en, zh);
            let mut herbivores: Vec<String> = Vec::new();

            // Try list column format first (Phase 0-4)
            if let Some(list_col) = herbivores_list_col {
                if let Some(list_series) = list_col.get_as_series(idx) {
                    if let Ok(str_series) = list_series.str() {
                        for herb_opt in str_series.into_iter() {
                            if let Some(herb) = herb_opt {
                                let herb = herb.trim();
                                if !herb.is_empty() {
                                    herbivores.push(herb.to_string());
                                }
                            }
                        }
                    }
                }
            }

            // Fallback to pipe-separated string format (legacy)
            if herbivores.is_empty() {
                if let Some(str_col) = herbivores_str_col {
                    if let Some(herbivores_str) = str_col.get(idx) {
                        if !herbivores_str.is_empty() {
                            for herb in herbivores_str.split('|') {
                                let herb = herb.trim();
                                if !herb.is_empty() {
                                    herbivores.push(herb.to_string());
                                }
                            }
                        }
                    }
                }
            }

            if herbivores.is_empty() {
                continue;
            }

            *plant_pest_counts.entry(display_name.clone()).or_insert(0) += herbivores.len();

            for pest in herbivores {
                pest_to_plants
                    .entry(pest)
                    .or_insert_with(Vec::new)
                    .push(display_name.clone());
            }
        }
    }

    if pest_to_plants.is_empty() {
        return Ok(None);
    }

    // Deduplicate plants per pest (in case same pest listed multiple times)
    for plants in pest_to_plants.values_mut() {
        plants.sort();
        plants.dedup();
    }

    let total_unique_pests = pest_to_plants.len();

    // Top 10 pests by plant count
    let mut top_pests: Vec<TopPest> = pest_to_plants
        .iter()
        .map(|(pest, plants)| {
            let category = OrganismCategory::from_name(pest, organism_categories, Some(OrganismRole::Herbivore));
            TopPest {
                pest_name: pest.clone(),
                category: category.display_name().to_string(),
                plant_count: plants.len(),
                plants: plants.clone(),
            }
        })
        .collect();

    // Sort by plant count
    // Sort by plant count (descending), then by name (ascending) for deterministic ordering
    top_pests.sort_by(|a, b| {
        b.plant_count.cmp(&a.plant_count)
            .then_with(|| a.pest_name.cmp(&b.pest_name))
    });
    top_pests.truncate(10);

    // Most vulnerable plants (by pest count)
    let mut vulnerable_plants: Vec<VulnerablePlant> = plant_pest_counts
        .iter()
        .map(|(plant, count)| VulnerablePlant {
            plant_name: plant.clone(),
            pest_count: *count,
        })
        .collect();

    // Sort by pest count (descending), then by name for deterministic ordering
    vulnerable_plants.sort_by(|a, b| {
        b.pest_count.cmp(&a.pest_count)
            .then_with(|| a.plant_name.cmp(&b.plant_name))
    });

    Ok(Some(PestProfile {
        total_unique_pests,
        top_pests,
        vulnerable_plants,
    }))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_pest_analysis_with_shared() {
        let df = df! {
            "wfo_taxon_name" => &["Plant A", "Plant B", "Plant C"],
            "herbivores" => &["Aphid|Beetle", "Aphid|Moth", "Beetle"]
        }
        .unwrap();

        let categories = FxHashMap::default();
        let profile = analyze_guild_pests(&df, &categories).unwrap().unwrap();

        assert_eq!(profile.total_unique_pests, 3);
        assert_eq!(profile.top_pests.len(), 3);

        // Aphid should be top (2 plants)
        assert_eq!(profile.top_pests[0].pest_name, "Aphid");
        assert_eq!(profile.top_pests[0].plant_count, 2);
    }

    #[test]
    fn test_pest_analysis_no_shared() {
        let df = df! {
            "wfo_taxon_name" => &["Plant A", "Plant B"],
            "herbivores" => &["Pest1|Pest2", "Pest3|Pest4"]
        }
        .unwrap();

        let categories = FxHashMap::default();
        let profile = analyze_guild_pests(&df, &categories).unwrap().unwrap();

        assert_eq!(profile.total_unique_pests, 4);
        assert_eq!(profile.top_pests.len(), 4);
    }

    #[test]
    fn test_pest_analysis_empty() {
        let df = df! {
            "wfo_taxon_name" => &["Plant A", "Plant B"],
            "herbivores" => &["", ""]
        }
        .unwrap();

        let categories = FxHashMap::default();
        let profile = analyze_guild_pests(&df, &categories).unwrap();
        assert!(profile.is_none());
    }
}

File: shipley_checks/src/Stage_4/explanation/pest_analysis.R
#!/usr/bin/env Rscript
#
# PEST VULNERABILITY ANALYSIS FOR M1
#
# Analyzes herbivore pest profiles for guilds, identifying:
# - Shared pests (generalists attacking 2+ plants)
# - Top 10 pests by interaction count
# - Most vulnerable plants
#
# This module provides qualitative information about pest vulnerability
# that supplements M1 scoring (phylogenetic diversity).
#
# Rust reference: src/explanation/pest_analysis.rs

#' Analyze pest profile for a guild
#'
#' Extracts herbivore information from organisms_df and identifies generalist
#' pests and vulnerable plants.
#'
#' @param guild_plants Data frame with guild plant data
#' @param organisms_df Data frame with organism associations
#'
#' @return List with pest profile, or NULL if no pests
#'
analyze_guild_pests <- function(guild_plants,
                                organisms_df) {

  n_plants <- nrow(guild_plants)

  if (n_plants == 0) {
    return(NULL)
  }

  # Filter organisms_df to guild plants
  plant_ids <- guild_plants$wfo_taxon_id
  guild_organisms <- organisms_df %>% dplyr::filter(plant_wfo_id %in% plant_ids)

  if (nrow(guild_organisms) == 0) {
    return(NULL)
  }

  # Build plant ID → name mapping
  plant_names <- setNames(
    guild_plants$wfo_scientific_name,
    guild_plants$wfo_taxon_id
  )

  # Build pest-to-plants mapping and plant-to-pest-count mapping
  pest_to_plants <- list()
  plant_pest_counts <- list()

  for (i in seq_len(nrow(guild_organisms))) {
    row <- guild_organisms[i, ]
    plant_id <- row$plant_wfo_id
    plant_name <- plant_names[[plant_id]]

    herbivores <- row$herbivores
    if (is.list(herbivores)) herbivores <- herbivores[[1]]

    if (!is.null(herbivores) && length(herbivores) > 0) {
      # Count pests for this plant
      if (is.null(plant_pest_counts[[plant_name]])) {
        plant_pest_counts[[plant_name]] <- 0
      }
      plant_pest_counts[[plant_name]] <- plant_pest_counts[[plant_name]] + length(herbivores)

      # Map pests to plants
      for (pest in herbivores) {
        if (is.null(pest_to_plants[[pest]])) {
          pest_to_plants[[pest]] <- character()
        }
        pest_to_plants[[pest]] <- c(pest_to_plants[[pest]], plant_name)
      }
    }
  }

  if (length(pest_to_plants) == 0) {
    return(NULL)
  }

  # Deduplicate plants per pest
  for (pest in names(pest_to_plants)) {
    pest_to_plants[[pest]] <- unique(pest_to_plants[[pest]])
  }

  total_unique_pests <- length(pest_to_plants)

  # Identify shared pests (2+ plants)
  shared_pests <- data.frame(
    pest_name = character(),
    plant_count = integer(),
    plants = character(),
    stringsAsFactors = FALSE
  )

  for (pest in names(pest_to_plants)) {
    plants <- pest_to_plants[[pest]]
    if (length(plants) >= 2) {
      shared_pests <- rbind(shared_pests, data.frame(
        pest_name = pest,
        plant_count = length(plants),
        plants = paste(plants, collapse = ", "),
        stringsAsFactors = FALSE
      ))
    }
  }

  # Sort by plant_count descending, then pest_name ascending
  if (nrow(shared_pests) > 0) {
    shared_pests <- shared_pests[order(-shared_pests$plant_count, shared_pests$pest_name), ]
  }

  # Top 10 pests by plant count (even if only 1 plant)
  top_pests <- data.frame(
    pest_name = character(),
    plant_count = integer(),
    plants = character(),
    stringsAsFactors = FALSE
  )

  for (pest in names(pest_to_plants)) {
    plants <- pest_to_plants[[pest]]
    top_pests <- rbind(top_pests, data.frame(
      pest_name = pest,
      plant_count = length(plants),
      plants = paste(plants, collapse = ", "),
      stringsAsFactors = FALSE
    ))
  }

  # Sort by plant_count descending, then pest_name ascending
  if (nrow(top_pests) > 0) {
    top_pests <- top_pests[order(-top_pests$plant_count, top_pests$pest_name), ]
    top_pests <- head(top_pests, 10)
  }

  # Most vulnerable plants (by pest count)
  vulnerable_plants <- data.frame(
    plant_name = character(),
    pest_count = integer(),
    stringsAsFactors = FALSE
  )

  for (plant_name in names(plant_pest_counts)) {
    vulnerable_plants <- rbind(vulnerable_plants, data.frame(
      plant_name = plant_name,
      pest_count = plant_pest_counts[[plant_name]],
      stringsAsFactors = FALSE
    ))
  }

  # Sort by pest_count descending, then plant_name ascending
  if (nrow(vulnerable_plants) > 0) {
    vulnerable_plants <- vulnerable_plants[order(-vulnerable_plants$pest_count, vulnerable_plants$plant_name), ]
  }

  # Return pest profile
  list(
    total_unique_pests = total_unique_pests,
    shared_pests = shared_pests,
    top_pests = top_pests,
    vulnerable_plants = vulnerable_plants
  )
}
